### ISSUE TITLE
Win10pro, RemoteApp Connection closed always ~ 2 mins

### ISSUE THREAD

**Issue by alturismo** [kimmknight/remoteapptool#15]
Hi, may anyone has a hint why all remoteapp sessions are closing appr. after 2 minutes,
no matter if in use or idle ... even while in active usage the logoff just appears ...

i made some changes then to the remoteapp host options and setted values like 7200 seconds,
but they dont change anything, like before, ~ 2 mins and session is closing.

also playing the gpedit.msc policys (user and computer) make no difference, actually i guess they only affect the rdp sessions and not remoteapp sessions.

may a hint where to look or how to bypass this behaviour ?

thanks ahead

as note, normal rdp sessions are staying alive (full rdp session) as expected, only remoteapp sessions are affected

**Comment by alturismo** [comment]
small update, i just tested this now on another win10 mashine, same result

rdp session working fine, remoteapp session starting, kicking out after 2 mins ...

just as note, 2 different win10, same behaviour

**Comment by bmg002** [comment]
Pretty sure that is going to be a setting in a GPO that is causing that.  I have Windows 10 Pro here and can run the remote apps for hours without any issues.  I have it set to boot after a certain amount of inactivity, but I can't imagine why you'd set your remoteapps to boot you out after 2 minutes.
When you say you made changes to the remoteapp host options, are you SURE it was the remoteapp options and not the remote desktop options?

Another thing to check is logs.  Windows event logs (application and security) is what I'd look at.  Application log will tell you if the app crashed or not, security will tell you if the user logged out or not;.
My guess without having access to your system is that you have something set to boot the user out after 2 minutes.  May be 2 minutes of inactivity, may be 2 minutes total.
The other thing, if this is a domain joined computer, you may need to adjust the settings on the AD server that pushes out the GPO.  Your manual settings may be overwritten.
My last thought is are you updating the RemoteApp Host options on the server hosting the remote app OR on the Windows 10 client?  They NEED to be adjusted on the server.

**Comment by alturismo** [comment]
@bmg002 hi, thanks for the feedback

about your questions, i did not set it up to kick me out after 2 minutes, its just happening.

about remote gpo options, i sure did look into remote desktop options and all i could find about remoteapp, but specific time rules about remoteapp, there where nothing i could find, may a hint if there are some ?

about app crashing, no, as i even see the logoff screen for a second before the connection is gone

its not a domain joined computer, so there are no AD settings to adjust

and of course on the server, the client(s) are no remoteapp hosted computer(s)

my personal last guess is may a conflict with rdpwrapper ... i ll try without rdpwrapper if thts the reason. i use it cause we may have 2 simultan rdp sessions (woman and me)

**Comment by bmg002** [comment]
What version of windows is your server?
And seeing the "logoff" screen doesn't mean the app didn't crash.  If your settings are to logoff after all remoteapp apps are terminated, then your app may be crashing causing the logoff.  

My first step would be to check the logs.  See what is triggering the logoff.
That being said, if the RDPWrapper you are using is this one:
https://github.com/stascorp/rdpwrap

It actually has an "enhancement" for adding support for RemoteApps.  Some people indicate it works fine, but they have an open issue/enhancement for it so I expect it is buggy.  There also appears to be an RDP limit with RDPWrapper of 15 users.

If this is a Windows Server OS, I would try using the built in tools for setting up the remoteapp as well and see if that helps with it.  
My expectation though is something in the logs (application or security, possibly system) will tell you why you are having that logoff issue.  Depending on how the RDP wrapper works, it might not allow 2 people to start the same app or the app may not allow 2 copies of itself to be running on the same machine.

As for the settings, the ones you want to look for (I think) are "set time limit for activve remote desktop services sessions".  
Specific timeouts for remoteapps I have only seen from the GUI on Windows Server 2016 and 2019 (the 2 server versions I use) from server manager.  And if you are using that tool, you likely don't need RemoteAppTool.

So to summarize, my first step would be to check the log to see WHY there is a logoff (is it user initiated, it is all apps have been terminated, is it a licensing thing... what caused the logoff?).  Next, I'd remove any additional users and/or tools you have for remote connections.  Isolate the server so only YOU are connected to ensure it isn't booting you out due to another user connecting to that remoteapp session.

**Comment by alturismo** [comment]
@bmg002 thanks again

windows 10 pro
rdpwrapper, exactly, i use it for 2 simultan sessions which is working perfectly fine for normal rdp sessions.

why i say its not the app crashing, i tested 10 different apps, all with the same behaviour, nvm about that now, i think im one step further.

i can nail it down it depends on the client, linux xfreerdp ... all connections from there will end in the 2 minute logoff behaviour, either directly or via guacamole.
i can confirm that using the remote app connection from another windows mashine directly its working as expected without the forced logoff.

So in the end it prolly will be some sort of security issue between xfreerdp connection, not triggering properly ...

when i check the logs i actually cant find anything which would lead me to a conclusion.

what i read about remoteapp and 2 minutes is, there is a default rule it seems, and it looks like the linux client cant trigger it or better dont see the warning so u can tell it that u still working, and as there is no specific remoteapp session time rule in gpedit ... im out of ideas ;)

![overview](https://user-images.githubusercontent.com/8406490/79358535-9b1a1100-7f41-11ea-86cc-99a0b766af7b.PNG)

[export.txt](https://github.com/kimmknight/remoteapptool/files/4482231/export.txt)

**Comment by bmg002** [comment]
I am a little confused.  I thought you said it was happening with windows 10 clients and now you are saying it is happening with xfreerdp?

I think I need to take a step back in this.  With windows clients, there is no problem at all, but with linux clients, you are having issues?
I would say try another RDP client from linux.  I am using Ubuntu via WSL with xfce and xrdp so I can RDP into it from my machine, then using xfreerdp from freerdp2-x11.  I found that freerdp-x11 was buggy and would not connect to most of my remoteapps.  Using this, I have had no issues with remoteapps running for hours or day unless I put restrictions in on how long it can remain connected.  Do you get any errors or warnings in your xfreerdp logs prior to being logged out?

Now, going over to your logs, event ID 4634 is the culprit for sure.  That one means "session was terminated and no longer exists".  As there is no event ID 4647, it means this wasn't a user initiated logoff.  The account appears to be named "media", right?
As a thought - are the users of the remoteapp using the same account?  If so, this could be causing problems.  Each user of a machine should have a unique username.  Reusing usernames when remoting into a machine (remoteapps is a form of remoting into a machine) will cause issues.  Looking in your log, it looks like all of the connections are using the account "media".  I did not dig through the whole log, but just a portion of it and that is what I see...

**Comment by alturismo** [comment]
if u understood win10 clients, sorry, no, win10 client i usually wouldnt need remoteapp here.

and yes, i use freerdp2-x11

tested now with different username (unique login) as test on another win10 pro remoteapp server mashine, same result, i ll dig deeper, at least i have a start point.

**Comment by bmg002** [comment]
Do you have the ability to isolate the machine?  What I mean is set it up so ONLY you can connect to it.  This could help with testing.
On top of that, are you up to date with all linux and windows patches and software updates?
If you set up a remoteapp for some simple program (such as notepad), do you still have that 2 minute ttimeout?
I was testing  it on my machine, using WSL (so not a REAL Ubuntu setup, but similar) and the remote app was running for 10 minutes with no problems.

I am still leaning towards that rdpwrapper as being the problem.  I would try a new server without rdpwrapper and see if it has the same behavior or not.  I would try to reduce the amount of 3rd party tools in between the client and server and see if it is one of the middle apps or not.

On a side note, if you only have  2 people connecting to the machine, you should be able to handle that without rdpwrapper by using the admin console connection.  One user connects normally, the other one connects to the console.

**Comment by alturismo** [comment]
Hi, yes, its a private PC only used by me and my woman, and i tested also with 1 user only.
and yes, all systems up to date.
and yes, the apps are tested simple (notepad, 7zip idle, ... up to office, all the same.
when u tested, may i ask how your string looks like ? simple as this ? or do u use a extra ?
xfreerdp /u:username /p:password /app:"||notepad" /v:192.168.1.200
i tested several combinations like /sec:.... /cert-ignore and so on ...
thanks ahead for a info here.

i disabled rdpwrapper for testing purposes but also that didnt work out properly, but i ll fire up a fresh windows, update it, only install remoteapp and see what happens.
i ll report back if that makes a change to finally see if remoteapp or the clients are the issue.

about the side note, my intentions why using rdpwrapper
scenario, always both users are usually locally logged on (of course only 1 works at a time)
woman @home local on the PC, im @work and simultan rdp into my account.

the remoteapp feature is more or less a nice to have, i can work with a full rdp session without issues as described, remoteapp is a nice feature which im experimenting with due i started using more linux clients (laptop, may change main pc to linux and keep 1 windows for win apps).
and afaik i cant achieve the simultan usage without rdpwrapper.

thanks for your input.

**Comment by alturismo** [comment]
@bmg002 

ok, fresh win10pro, only made all available updates
enabled remotedesktop, changed remotedesktop settings to allow other connections (not only NLA)
insatlled remoteapp, added Notepad (Editor) as app,  result is the same, so its not rdpwrapper related i can say.

after some more reading, i found a entry in the guacamole forums that there was a change in xfreerdp2 which is currently used and that the "PDU" handshake is not in the rail by default anymore,
which they will fix in their 1.2 version, but as its working fine at your side with freerdp im a bit confused ;) cause then this wouldnt be the issue.

Nevermind now, if u may would give me a sample from your xfreerdp2 cmd string u use i d like to give it a try in case u use a special setting.

i also can say remoteapp works flawlessly with other win clients using the rdp file from remoteapp, left a vlc remoteapp testrun on, its running for hours ...

**Comment by alturismo** [comment]
@bmg002 

the link to the issue from guacamole how i come that its an freerdp2 issue

https://issues.apache.org/jira/browse/GUACAMOLE-978

**Comment by bmg002** [comment]
My command was similar but different.  I included a domain (as my computer is domain joined) and I used an RDP file rather than the app name and server IP.
I used RemioteAppTool to create the RDP file, copied that over to a location visible and easy to find on the linux side and ran:
xfreerdp /u:username /d:domain rdpfilename.rdp

I never ever ever put my password somewhere where it can be viewed like that as it is in the command history then and anyone with root access to the linux machine can then see my password.  My personal policy is if the password is visible to me, it is visible to the world and therefore I should not type my password into that field.

**Comment by MrBrianGale** [comment]
Also, what version of xffreerdp are you using (xfreerdp /version)?
Mine comes back as:
This is FreeRDP version 2.0.0-dev5 (n/a)

Could be we are running different versions of xfreerdp?

**Comment by alturismo** [comment]
sadly not ;)

alturismo@AlsSurface:~$ xfreerdp --version
This is FreeRDP version 2.0.0-dev5 (2693389a+debian)
alturismo@AlsSurface:~$

**Comment by MrBrianGale** [comment]
Looks like a SLIGHTLY different version as mine came back as (n/a) at the end and yours was 2693389a+debian... but we are both on 2.0.0-dev5... May just need to wait for the next freerdp2 release?  Or possibly try freerdp-x11 instead of freerdp2-x11?

**Comment by MrBrianGale** [comment]
With it working on my end it COULD be that I am running Ubuntu via WSL.  COULD be that WSL is doing some magic and fixing things on its own.  Or could be my Ubuntu version for WSL is doing some magic and fixing xfreerdp?  Or xfce fixing stuff... 
OR, grasping at straws, MAYBE it is using an RDP file instead of the argument?  Maybe when it loads an .rdp file it pushes across extra data for the handshake?  I could test the rdp file thing pretty easily, but WSL is a bit more difficult to test on my end unfortunately.

**Comment by alturismo** [comment]
thanks alot for trying to help, im waiting now for a freerdp upgrade and/or guacamole upgrade, too lazy now to build the 1.2 staging release myself.

i know it has been working when i played with it before, but that was some time ago ... so pretty sure before freerdp2 was out.

thanks again, i ll report back if that was(is the reason

**Comment by MrBrianGale** [comment]
Looking at guacamole, it looks like they have fixed things on their end.  the bug you linked above is marked as clsoed-fixed.

Are you still having the issue?  If so, it might not hurt to try running it against the older freerdp - freerdp-x11 instead of freerdp2-x11 just to see if that helps?

**Comment by alturismo** [comment]
hi, it is marked as fix and will be implemented in guac 1.2 which is not releases yet.

i did a fresh install on my laptop of ubuntu 20 and made a fast test, here even no remoteapp at all is working, so i wait for guac 1.2 and report back.

about the -dev... i wonder too, but i installed simple <apt install freerdp2-x11 and its the same version.

i report back when i figured it.

**Comment by alturismo** [comment]
@MrBrianGale hi, updated today to guacamole 1.2 (where the fix should be included) and yes, working now

before i always had the ~ 2 min disconnect ;)
![image](https://user-images.githubusercontent.com/8406490/86223656-1cb82d00-bb88-11ea-83fc-1dad8b6f963b.png)

**Comment by MrBrianGale** [comment]
I am glad to hear it is working now!
I will close this issue.
